lib/Dialect/HW/HWOps.cpp

```diff
@@ -889,16 +892,19 @@ static ParseResult parseArrayConcatTypes(OpAsmParser &p,
   Type elemType;
   uint64_t resultSize = 0;
   do {
-    ArrayType ty;
+    Type ty;
     if (p.parseType(ty))
+      return p.emitError(p.getCurrentLocation(), "Expected type");
+    if (!type_isa<ArrayType>(ty))
```

Comment:
I'd suggest creating a `type_dyn_cast` (maybe just converting `type_cast` over) and using it here and above.

Reply:
Fair, I didn't see these use cases at first, but a dyn_cast would be appropriate here. I'll keep all three--there are some places where ODS guarantees the type_cast will succeed and we shouldn't need to use a dyn_cast.